File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        keyMap = {
            '(': ')',
            '{': '}',
            '[': ']'
        }
        
        stack = []
        
        for ch in s:
            if ch in keyMap:
                stack.append(ch)
            elif len(stack) == 0 or ch != keyMap[stack.pop()]:
                return False

            
        return len(stack) == 0

#         stack = [] # only use append and pop
#         pairs = {
#             '(': ')',
#             '{': '}',
#             '[': ']'
#         }
#         for bracket in s:
#             if bracket in pairs:
#                 stack.append(bracket)
#             elif len(stack) == 0 or bracket != pairs[stack.pop()]:
#                 return False

#         return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> tracker = new Stack<>();
        
        HashMap<Character,Character> map = new HashMap<>();
        
        map.put('{','}');
        map.put('[',']');
        map.put('(',')');
        
        for (char ch : s.toCharArray()){
            if(map.containsKey(ch)){
                tracker.push(ch);
            }
            else if(tracker.isEmpty()) {
                return false;
            }
            else{
                if(ch != map.get(tracker.pop())){
                    return false;
                }
            }
        }
        
        return tracker.isEmpty();
    }
}